### ISSUE TITLE
Export report-plist to communicate status from script to app

### ISSUE THREAD

**Issue by homebysix** [homebysix/recipe-robot#45]
After the Recipe Robot script finishes, it should output a report-plist to /tmp or the Cache folder, so that the planned Mac app can read the status of the last run, any errors, and other information. Once this is done we should be able to go full speed ahead on the Mac app.

**Comment by homebysix** [comment]
I took a shot at this in the [output-plist](https://github.com/homebysix/recipe-robot/tree/output-plist) branch. @eahrold, what else does the output plist need in order for the app to accurately display the run results?

@sheagcraig The variables and functions might need some Pythonizing. `report` is a global variable right now, and I know there's got to be a better way. Also, the imports are a bit messier now because the function needs to get passed back and forth between modules. You'd be better at cleaning that up than me.

**Comment by eahrold** [comment]
Not end of run related, but we may need to wrap back and have a non ASCII color progress. I'm not sure how much you want to display in the GUI but I currently have it showing -v.

**Comment by homebysix** [comment]
`-v` is perfect for GUI progress.

Not sure what you mean about ASCII color. Is the color change going to mess something up in the app progress display?

**Comment by eahrold** [comment]
Sorry ANSI colors. Brain got twisted.

**Comment by homebysix** [comment]
Oh, I know which colors you meant — I'm just unclear about how that affects the app?

**Comment by sheagcraig** [comment]
Yep. Fixing the report global.

I'm guessing there's a more proper place than `/tmp/report.plist` also.

Python has a tempfile module, but you have to be careful-if you use the wrong one it deletes the file when the file handle gets closed. But it does make a randomized subfolder in tmp.

However, I think it might make more sense to put it in App Support where you don't have to worry about it  interacting with the tmp folder. Should it have a unique, timestamped name?

I don't see how this report.plist is getting used in the app? I guess it isn't yet?

**Comment by sheagcraig** [comment]
So I think _one_ way to do this would be to run a basic http server on an open port. It's pretty simple to implement a simple REST API. Then the app could poll recipe robot for status periodically.

The only problem I foresee with this approach is that with the new push towards all apps using App Transport Security, we're back into python TLS hell. It's not required yet, but communication between an app and its backend will be required to use HTTPS, and it sounds like higher level Cocoa API's require TLS.

**Comment by sheagcraig** [comment]
I've got the `facts` outputting as the report now. Not super useful for ongoing status yet-I'll investigate writing to the `/tmp/report.plist` as things change rather than just at the end as the next step.

**Comment by eahrold** [comment]
I'd vote against the App Support dir, and suggest use the user's tmp directory instead.

```Python
import tempfile
print tempfile.gettempdir()
```

**Comment by sheagcraig** [comment]
So check it out. I made a branch called facts-writing where I made a Facts class that is basically a dict that writes to a plist every time a value is changed.

I have no idea what the performance of that would be like; we can profile it later if it seems like a good idea to use; I'm sure it could cache stuff and only write periodically or something.

I was just trying to get something for the app to read.

So now ALL of the facts and args are updated as recipe-robot goes for the app's use. I was thinking it would be cool to have a bunch of little status lights that would blink on as certain facts were determined. Like one for icon one for sparkle feed url, etc. It would potentially be a little more interesting than just watching a textbox scroll by.

If you think this is a good idea, let me know and I can spend some time researching how best to do this so it doesn't doing a ton of unnecessary IO, although honestly, the resulting total facts dict is not that large. I'm sure it probably isn't even noticeable. But it's for style points ;)

**Comment by eahrold** [comment]
This sounds awesome. And all in all rr isn't a performance product so I wouldn't worry about io much. 

The only thing I'm not crazy about is writing to a file which would require feventing. Another possible approach is to have an output arg that would change the behavior of data to stdout disabling 'logging' and instead would send json to stdout for each fact or an acculimination.
We could then simply serialize that as it comes in.

**Comment by sheagcraig** [comment]
You know, one slightly alternate method would be to go ahead and switch over to use the PyObjC stuff for preferences. Then, when a fact gets set, instead of writing the whole plist, the code could just set that key with CFPreferencesSetValue (because that allows you to write outside of the app's domain) and then CFPreferencesSynchronize. This would only cause the new pref to get written each time.

Perhaps part of that override of the Facts __setitem__ could add in some kind of event triggering or callback. So the app could run recipe-robot with an option that specifies a callback that it knows about. When the RR script thinks it has done something that the app should know about, it runs the callback, which basically says "Hey! Something happened! Check the facts plist! And I imagine it could contain a list as an arg to the callback that has all of the changed keys, so the APP would know exactly where to look.

**Comment by sheagcraig** [comment]
It wouldn't probably be NSEvent that we would use... That seems to be for UI activity.

**Comment by sheagcraig** [comment]
CFNotification? NSNotification?

**Comment by sheagcraig** [comment]
CFNotificationGetDistributedCenter allows you to send notifications to different apps.

Arguably, this obviates the need to write the Facts out in the first place. I think if we made Facts a Cocoa object, we could post a notification from RR with the Facts object as the notification sender parameter.

Something to toy around with.

**Comment by eahrold** [comment]
Good idea.

**Comment by eahrold** [comment]
NSDistributedNotifications would work well. I think that's how munki use to do it, maybe it still does.

**Comment by eahrold** [comment]
So I set up a NotificationListener class in the app that looks like this

``` Swift
enum NoteType: Int {
    case Error, Reminders, Warnings, Recipes, Icons

    static func fromName(name: String) -> NoteType? {
        for t in self.all() {
            if t.name == name {
                return t
            }
        }
        return nil
    }

    var name: String {
        switch self {
        case Error: return "com.elliotjordan.recipe-robot.dnc.error"
        case Reminders: return "com.elliotjordan.recipe-robot.dnc.reminders"
        case Warnings: return "com.elliotjordan.recipe-robot.dnc.warnings"
        case Recipes: return "com.elliotjordan.recipe-robot.dnc.recipes"
        case Icons: return "com.elliotjordan.recipe-robot.dnc.icons"
        }
    }

    static func all()-> [NoteType]{
        var all = [NoteType]()
        for idx in NoteType.Error.rawValue...NoteType.Icons.rawValue {
            if let noteType = NoteType(rawValue: idx){
                all.append(noteType)
            }
        }
        return all
    }
}

class NotificationListener: NSObject {
    var notificationHandler: ((name: NoteType, info: [String:AnyObject]) -> (Void))?
    private let center = NSDistributedNotificationCenter.defaultCenter()

    override init(){
        super.init()

        for n in NoteType.all() {
            center.addObserver(self, selector: "noticed:",
                name: n.name,
                object: nil)
        }
    }

    init(noteTypes: [NoteType]) {
        super.init()
        for n in noteTypes {
            center.addObserver(self, selector: "noticed:",
                name: n.name,
                object: nil)
        }
    }

    deinit {
        NSDistributedNotificationCenter.defaultCenter().removeObserver(self)
    }

    func noticed(note: NSNotification){
        if notificationHandler != nil{
            if let noteType = NoteType.fromName(note.name), dict = note.userInfo as? [String: AnyObject] {
                notificationHandler!(name: noteType, info: dict)
            }
        }
    }
}
```

And you can send to it like this it like this
```Python
import Foundation
from AppKit import *

userInfo = {"message": "error message"}
center = Foundation.NSDistributedNotificationCenter.defaultCenter()
center.postNotificationName_object_userInfo_options_(
        "com.elliotjordan.recipe-robot.dnc.error",
        None,
        userInfo,
        NSNotificationDeliverImmediately)
```

What I like about this approach is we'll be able to use a switch statement to handle the UI updates.

```Swift
 listener.notificationHandler = {
            noteType, info in
            switch noteType {
            case .Error:
                // do something
                break
            case .Reminders:
                break
                // do something else
            default:
                break
            }
        }
```

How's this sound to you?

**Comment by sheagcraig** [comment]
That's awesome. I'll work from that. I was nervous about writing both sides!

Finishing up some exception handling, then this.

**Comment by sheagcraig** [comment]
As it is going now, I was writing the entire Facts object out every time something changed.

Should we add another value to the enum for just "Information" or something like that? I can of course only send notifications for the above values, but we've got more data to share!

**Comment by sheagcraig** [comment]
Also, it was nice to have all of this data written out, but I guess RR does it when you do the debug mode too? It may be something we want to do on the way out-write Facts to a plist/log file somewhere.

**Comment by sheagcraig** [comment]
@eahrold I set everything up for Facts keys errors, reminders, warnings, to fire notifications, and for errors to quit.

All, I still need to switch over all of the robo-printing to-instead setting facts before I can merge this in.

AND I just realized that recipes and icons need to have the notification-firing behavior... I'll do that right now.

**Comment by sheagcraig** [comment]
This is all in the "facts-writing" branch. @homebysix heads up-this means that all you have to do in the main code now to print an error, reminder, or warning, is to append an item to the facts["errors"] (or "reminders" or "warnings") lists, and the overriden lists will handle notifying the app and exiting, if needed.

**Comment by eahrold** [comment]
@sheagcraig I'll merge into the dmc-and-reface branch and test it out.
For me the way I'm thinking about it now is that each notification should be a single self contained message used for progress updates. I'm not sure if this is what you were thinking, but I do like the idea of sending a final "Information" notification with the accumulated dict for all the values at the end of the run.

**Comment by eahrold** [comment]
Hey Shea, I just pulled in the facts branch and started seeing this.

```
$ ./recipe-robot -v /Applications/Deeper.app


### ... ###
                
Processing /Applications/Deeper.app ...
Input path looks like an app.
Validating app...
    App seems valid
Getting app name...
    App name is: Deeper
Getting bundle identifier...
    Bundle idenfitier is: com.titanium.Deeper
Checking for a Sparkle feed...
    No Sparkle feed
Determining whether app was downloaded from the Mac App Store...
    App did not come from the App Store
Looking for version key...
    Version key is: CFBundleShortVersionString (1.7.0)
Looking for app icon...
    App icon is: /Applications/Deeper.app/Contents/Resources/Deeper
Getting app description from MacUpdate...
    Description: Enable hidden features in OS X.
Gathering code signature information...
    App is not signed
[ERROR] Recipe Robot exploded with unexpected error: global name 'recipes' is not defined
[ERROR] Traceback (most recent call last):
[ERROR]   File "./recipe-robot", line 1904, in main
[ERROR]     time, _ = recipe_robot_lib.generate_recipes(facts, prefs, recipes)
[ERROR] NameError: global name 'recipes' is not defined
[ERROR] global name 'recipes' is not defined
```

**Comment by sheagcraig** [comment]
Yeah-I've been working in 30 minute bursts during the baby's naps. It's not really merge able yet. I have a commit just ready to go sitting on the laptop that should be good to go. I'll push it as soon as I get to work.

**Comment by eahrold** [comment]
That sounds all too familiar. I've got two so mine's more like 5 when I magically get an nap overlap.

But no problem, I've sorted it out on my end for testing, but I'll just discard those changes when you push what you've got.

**Comment by sheagcraig** [comment]
Okay. In.

**Comment by sheagcraig** [comment]
Also-I fixed the mistake in the Facts setitem.

**Comment by sheagcraig** [comment]
I think we're done with this, right?

**Comment by homebysix** [comment]
Yep!
